src/sos/objectstorage.go

```diff
@@ -0,0 +1,124 @@
+package sos
+
+import (
+	"context"
+	"fmt"
+	pipeline "github.com/ccremer/go-command-pipeline"
+	egoscale "github.com/exoscale/egoscale/v2"
+	"github.com/vshn/exoscale-metrics-collector/src/kubernetes"
+	exoscalev1 "github.com/vshn/provider-exoscale/apis/exoscale/v1"
+	"golang.org/x/exp/slices"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+var (
+	organizationLabel = "appuio.io/organization"
+	namespaceLabel    = "crossplane.io/claim-namespace"
+)
+
+type ObjectStorage struct {
+	k8sClusters    *[]kubernetes.Cluster
+	exoscaleClient *egoscale.Client
+	bucketDetails  []BucketDetail
+}
+
+type BucketDetail struct {
+	ClusterName  string
+	Namespace    string
+	Organization string
+	BucketName   string
+}
+
+func NewObjectStorage(exoscaleClient *egoscale.Client, k8sClusters *[]kubernetes.Cluster) ObjectStorage {
+	return ObjectStorage{
+		exoscaleClient: exoscaleClient,
+		k8sClusters:    k8sClusters,
+	}
+}
+
+func (o *ObjectStorage) Execute(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Running metrics collector by step")
+
+	p := pipeline.NewPipeline[context.Context]()
+	p.WithSteps(
+		p.NewStep("Fetch managed buckets", o.fetchManagedBuckets),
+		p.NewStep("Get bucket usage", o.getBucketUsage),
+	)
+	return p.RunWithContext(ctx)
+}
+
+func (o *ObjectStorage) getBucketUsage(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Fetching bucket usage from Exoscale")
+	resp, err := o.exoscaleClient.ListSosBucketsUsageWithResponse(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, bucketUsage := range *resp.JSON200.SosBucketsUsage {
+		name := *bucketUsage.Name
+		zoneName := *bucketUsage.ZoneName
+		size := *bucketUsage.Size
+		createdAt := *bucketUsage.CreatedAt
+
+		log.V(1).Info("Trying to find a match in a cluster", "bucket name", name)
+		// Match bucket name from exoscale with the bucket found in one of the clusters
+		bucketIndex := slices.IndexFunc(o.bucketDetails, func(ni BucketDetail) bool { return ni.BucketName == name })
+
+		if bucketIndex == -1 {
+			log.Info("Cannot find bucket in any cluster", "bucket name", name)
+			continue
+		}
```

Comment:
We can also avoid a library here with a simple for-implementation by ourselves. Probably better to introduce our own function here that is more descriptive than `indexfunc`

Reply:
It will be removed with the next PR anyway. I will leave it as is for now.